File: source/downloads/code/LeetCode/Java/Word-Ladder.java
public class Solution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {

	HashSet<String> visited = new HashSet<String>();

	visited.add(beginWord);
	wordList.add(endWord);
	int distance = 1;
	while (!visited.contains(endWord)){
	    Set<String> toAdd = new HashSet<String>();
	    for (String each : visited){
		// All strings in visited
		for(int i=0; i<each.length(); i++){
		    char[] chars = each.toCharArray();
		    for (char ch = 'a'; ch <= 'z'; ch++){
			chars[i] = ch;
			String word = new String(chars);
			// find matched
			if(wordList.contains(word)){
			    toAdd.add(word);
			    wordList.remove(word);
			}
		    }
		}
	    }
	    distance ++;
	    if(toAdd.size()==0) return 0;
	    visited = toAdd;
	}
	return distance;
    }
}

File: downloads/code/LeetCode/Python/Word-Ladder.py
class Solution(object):
    def ladderLength(self, beginWord, endWord, wordList):
        """
        :type beginWord: str
        :type endWord: str
        :type wordList: Set[str]
        :rtype: int
        """
        alpha = "abcdefghijklmnopqrstuvwxyz"
        n = len(beginWord)
        stack = []
        stack.append([beginWord,1])
        wordList.add(endWord)
        while stack:
            tmp = stack.pop(0)
            word = tmp[0]
            val = tmp[1]
            alpha = "abcdefghijklmnopqrstuvwxyz"
            if word == endWord:
                return val

            for i in range(len(word)):
                for j in alpha:
                    if j == word[i]:
                        continue
                    newWord = word[:i]+j+word[i+1:]
                    if newWord in wordList:
                        stack.append([newWord,val+1])
                        wordList.remove(newWord)
        return 0
